### PR TITLE
gateway: make IPNSHostname work on responses too

### PR DESCRIPTION
**Pull request by lgierth**
IPNSHostnameOption() touches the URL path only on the way in,
but not on the way out. This commit makes it complete by
touching the following URLs in responses:

- Heading, file links, back links in directory listings
- Redirecting /foo to /foo/ if there's an index.html link
- Omit Suborigin header

License: MIT
Signed-off-by: Lars Gierth <larsg@systemli.org>

### CODE REVIEW

**Comment by jbenet** [comment]
@lgierth i think i get it, but to be sure, can maybe describe a bit more in the code as a comment somewhere? or maybe want a test case that shows the cases it's supposed to be manipulating?

**Comment by lgierth** [comment]
I clarified the comments.

> maybe want a test case that shows the cases it's supposed to be manipulating?

Yes agreed, if we start serving our website using this, we better make sure it does what it says on the tin.

**Comment by jbenet** [comment]
Yep. :+1:



—

File: core/corehttp/gateway_test.go [hunk 1]
```diff
@@ -130,3 +139,64 @@ func TestGatewayGet(t *testing.T) {
...
+	}
+
+	k, err := dagn1.Key()
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Logf("k: %s\n", k)
+	ns["/ipns/example.net"] = path.FromString("/ipfs/" + k.String())
+
+	req, err := http.NewRequest("GET", ts.URL+"/ipfs/"+k.String(), nil)
```
Comment (lgierth) [inline core/corehttp/gateway_test.go]: The responds with 404...

File: core/corehttp/gateway_handler.go [hunk 2]
```diff
@@ -112,10 +123,17 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
...
 	// Suborigin header, sandboxes apps from each other in the browser (even
-	// though they are served from the same gateway domain). NOTE: This is not
-	// yet widely supported by browsers.
-	pathRoot := strings.SplitN(urlPath, "/", 4)[2]
-	w.Header().Set("Suborigin", pathRoot)
+	// though they are served from the same gateway domain).
+	//
+	// Omited if the path was treated by IPNSHostnameOption(), for example
+	// a request for http://example.net/ would be changed to /ipns/example.net/,
+	// which would turn into an incorrect Suborigin: example.net header.
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: maybe it should be corrected then? instead of omitted?

**Comment by lgierth** [comment]
There are a few unneccessary comments and a bit of duplication in there, I can clean that up, or we leave it for the great gateway bonfire (go-ipfs-gateway).

File: core/corehttp/gateway_handler.go [hunk 3]
```diff
@@ -205,9 +228,17 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 				}
 			}
 
+			// strip /ipfs/$hash from backlink if IPNSHostnameOption touched the path.
+			// we could use originalUrlPath like we do with the redirect,
+			// but i don't wanna break the switch-case right above, not now.
+			if urlPath != originalUrlPath {
+				backLink = "/" + strings.Join(pathSplit[4:], "/")
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: is that the only case where `urlPath != originalUrlPath`? because if not, we might panic the process (by accessing `pathSplit[4:]`. 

Maybe:
```go
if urlPath != originalUrlPath && pathSplit[1] == "ipns" {
  backLink = "/" + strings.Join(pathSplit[4:], "/")
}
```
also, shouldn't it be `pathSplit[3:]` (3) ?

**Comment by jbenet** [comment]
> There are a few unneccessary comments and a bit of duplication in there, I can clean that up, or we leave it for the great gateway bonfire (go-ipfs-gateway).

yeah should be fine

**Comment by jbenet** [comment]
@lgierth ok, CRed o/ -- a few comments

File: core/corehttp/gateway_handler.go (hunk 2 above)
Comment (lgierth) [inline core/corehttp/gateway_handler.go]: With IPNSHostname the gateway serves one object per hostname, while otherwise it serves all the objects that exist (through /ipfs and /ipns). I thought there'd be no need to separate it from anything, while under /ipfs and /ipns, each object and identity should be separated.

File: core/corehttp/gateway_handler.go (hunk 3 above)
Comment (lgierth) [inline core/corehttp/gateway_handler.go]: Yeah I'm piggybacking "path was touched by IPNSHostname" here, will change.

File: core/corehttp/gateway_handler.go [hunk 4]
```diff
@@ -90,6 +90,17 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
...
 
+	// IPNSHostnameOption might have constructed an IPNS path using the Host header.
+	// In this case, we need the original path for constructing redirects
+	// and links that match the requested URL.
+	// For example, http://example.net would become /ipns/example.net, and
+	// the redirects and links would end up as http://example.net/ipns/example.net
+	originalUrlPath := urlPath
+	hdr := r.Header["X-IPNS-Original-Path"]
+	if len(hdr) > 0 {
+		originalUrlPath = hdr[0]
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: maybe want to set here: `ipnsHostRedirection = true` and use that for the conditions below?

File: core/corehttp/gateway_handler.go (hunk 2 above)
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: ohhhhh that makes sense.

File: core/corehttp/gateway_handler.go (hunk 4 above)
Comment (lgierth) [inline core/corehttp/gateway_handler.go]: Yeah I introduced a variable in order to avoid the piggybacking. I also fixed backlink edge-cases and added more tests for them.

It should be good to go now.

**Comment by jbenet** [comment]
@lgierth LGTM! one small comment above o/, and this:
- [ ] if tests dont pass on all commits, squash the non-passing ones.

**Comment by lgierth** [comment]
Here we go! I don't know if any of the previous commits failed, but I squashed them regardless.

That Travis failure is from my jumping between lgierth:gateway-host-header and ipfs:gateway-host-header...

**Comment by whyrusleeping** [comment]
looks like one of travises OSX machines failed pretty badly. Other than that, the tests look fine.

**Comment by jbenet** [comment]
This LGTM!
